xde/src/xde.rs

```diff
@@ -268,6 +277,25 @@ struct XdeDev {
 unsafe extern "C" fn _init() -> c_int {
     xde_devs.init(KRwLockType::Driver);
     mac::mac_init_ops(&mut xde_devops, XDE_STR);
+
+    // By virtue of being a MAC provider, we're a STREAMS driver as
```

Comment:
> `mac_init_ops` does allow passing `NULL` for `dev_ops` for providers to fill in themselves (softmac for instance does this). I considered that but that also meant doing a lot more when all we want is `ENOSTR` returned sometimes.

`mac_init_ops` allows it strictly speaking, but that's kind of a private thing for `softmac`. That is, what we document for folks is that they have to fill it in. I'll brainstorm on this more. Let me know if you end up having any ideas. The biggest gotcha is wanting it independently of `mac_register()`. But I'll need to better familiarize myself with some of the flow here.

Reply:
Making this change in dld_str.c would suffice and wouldn't require any API changes to `mac_init_ops` or such.
```diff
diff --git a/usr/src/uts/common/io/dld/dld_str.c b/usr/src/uts/common/io/dld/dld_str.c
index 9f89165455..1b9932a9ec 100644
--- a/usr/src/uts/common/io/dld/dld_str.c
+++ b/usr/src/uts/common/io/dld/dld_str.c
@@ -272,20 +272,33 @@ int
 dld_str_open(queue_t *rq, dev_t *devp, void *private)
 {
 	dld_str_t	*dsp;
 	major_t		major;
 	minor_t		minor;
 	int		err;
 
 	major = getmajor(*devp);
 	minor = getminor(*devp);
 
+	/*
+	 * Half the 32-bit minor space is reserved for private use by the driver so
+	 * we bail out here with `ENOSTR` to indicate specfs should retry the open
+	 * with the driver's character based `open(9E)`. For a typical STREAMS
+	 * driver, that would just be `nodev` which would simply return `ENODEV`.
+	 * But a dual-personality device can choose to implement the character
+	 * based `open(9E)` for some minor nodes. A driver wanting a separate
+	 * STREAMS interface altogether would presumably have already provided its
+	 * own `streamtab`.
+	 */
+	if (minor >= mac_private_minor())
+		return (ENOSTR);
+
 	/*
 	 * Create a new dld_str_t for the stream. This will grab a new minor
 	 * number that will be handed back in the cloned dev_t.  Creation may
 	 * fail if we can't allocate the dummy mblk used for flow-control.
 	 */
 	dsp = dld_str_create(rq, DLD_DLPI, major,
 	    ((minor == 0) ? DL_STYLE2 : DL_STYLE1));
 	if (dsp == NULL)
 		return (ENOSR);
 ```

For any existing case where we ended up here with `minor >= mac_private_minor()` it would just fail in `dld_str_create` with `ENOENT`. With the change it would now bounce to`devo_cb_ops->cb_open` which, assuming a standard STREAMS driver, would be `nodev()` and return `ENODEV`. Though, I feel like hitting this case today seems like a bug in the driver as it presumably created the minor but never hooked up some way of handling it?

Reply:
Created upstream bug for this: https://www.illumos.org/issues/15342. I'll also put up that change and solicit some review/feedback.

Reply:
ab5fdce2f303677d8d591f63477aca9fa6dcdaa9 removes the open routine swapping hack when paired together with https://code.illumos.org/c/illumos-gate/+/2608